File: lib/tabs/tabs.dart
import "package:angular2/angular2.dart";
import 'package:ng2_strap/common.dart';
import 'package:node_shims/js.dart';
// todo: add active event to tab

// todo: fix? mixing static and dynamic tabs position tabs in order of creation
@Component (
    selector: "tabset", inputs: const [ "vertical", "justified", "type"],
    template: '''
    <ul class="nav"
        [ngClass]="{
          'nav-stacked' : vertical,
          'nav-justified' : justified,
          'nav-tabs' : type == 'tabs',
          'nav-pills' : type == 'pills'
        }"
        (click)="\$event.preventDefault()">
        <li *ngFor="#tabz of tabs" class="nav-item" [ngClass]="{active: tabz.active, disabled: tabz.disabled}">
          <a href class="nav-link" [ngClass]="{active: tabz.active, disabled: tabz.disabled}" (click)="tabz.active = true">
            <span [ng-transclude]="tabz.headingRef">{{tabz.heading}}</span>
          </a>
        </li>
    </ul>
    <div class="tab-content">
      <ng-content></ng-content>
    </div>
  ''', directives: const [CORE_DIRECTIVES, NgClass, NgTransclude])
class Tabset implements OnInit {
  bool vertical = false;

  bool justified = false;

  String type;

  List<Tab> tabs = [];

  Tabset();

  ngOnInit() {
    type ??= "tabs";
  }

  addTab(Tab tab) {
    tabs.add(tab);
    tab.active = tabs.length == 1 && tab.active != false;
  }

  removeTab(Tab tab) {
    var index = tabs.indexOf(tab);
    if (identical(index, -1)) {
      return;
    }
    // Select a new tab if the tab to be removed is selected and not destroyed
    if (tab.active && tabs.length > 1) {
      // If this is the last tab, select the previous tab. else, the next tab.
      var newActiveIndex = identical(index, tabs.length - 1)
          ? index - 1
          : index + 1;
      tabs [ newActiveIndex ].active = true;
    }
    slice(tabs, index, 1);
  }
}
// TODO: templateUrl?
@Directive (selector: "tab, [tab]",
    inputs: const [ "active", "disable", "disabled", "heading"],
    outputs: const [ "select", "deselect"],
    host: const { "[class.tab-pane]" : "true", "[class.active]" : "active"})
class Tab implements OnInit, OnDestroy, DoCheck {
  Tabset tabset;

  bool _active = true;

  bool disabled = false;

  String heading;

  TemplateRef headingRef;

  EventEmitter select = new EventEmitter ();

  EventEmitter deselect = new EventEmitter ();

  Tab(this.tabset) {
    tabset.addTab(this);
  }

  @deprecated
  set disable(bool v) {
    print("DEPRECATED use `disabled` property (not `disable`)");
    disabled = v;
  }

  /** DEPRECATE disable */
  @deprecated
  get disable {
    return disabled;
  }

  /** tab active state toogle */
  get active {
    return _active;
  }

  set active(bool active) {
    active ??= true;
    if (disabled && active != null || !active) {
      if (!active) {
        _active = active;
      }
      deselect.add(this);
      return;
    }
    _active = active;
    select.add(this);
    tabset.tabs.forEach((Tab tab) {
      if (!identical(tab, this)) {
        tab.active = false;
      }
    });
  }

  bool ngDoCheck() {
    return true;
  }

  ngOnInit() {}

  ngOnDestroy() {
    tabset.removeTab(this);
  }
}

@Directive (selector: "[tab-heading]")
class TabHeading {
  TemplateRef templateRef;

  TabHeading(this.templateRef, Tab tab) {
    tab.headingRef = templateRef;
  }
}

const TABS_DIRECTIVES = const [Tab, TabHeading, Tabset];

File: components/tabs/tabs.ts
import {
  Component, Directive,
  OnInit, OnDestroy, DoCheck,
  Input, Output, HostListener, HostBinding,
  TemplateRef, EventEmitter
} from 'angular2/core';
import { NgClass } from 'angular2/common';
import { NgTransclude, IAttribute } from '../common';

// todo: add active event to tab
// todo: fix? mixing static and dynamic tabs position tabs in order of creation
@Component({
  selector: 'tabset',
  directives: [NgClass, NgTransclude],
  template: `
    <ul class="nav" [ngClass]="classMap" (click)="$event.preventDefault()">
        <li *ngFor="#tabz of tabs" class="nav-item" [ngClass]="{active: tabz.active, disabled: tabz.disabled}">
          <a href class="nav-link" [ngClass]="{active: tabz.active, disabled: tabz.disabled}" (click)="tabz.active = true">
            <span [ngTransclude]="tabz.headingRef">{{tabz.heading}}</span>
          </a>
        </li>
    </ul>
    <div class="tab-content">
      <ng-content></ng-content>
    </div>
  `
})
export class Tabset implements OnInit {
  @Input() private get vertical() {
    return this._vertical;
  };

  @Input() private get justified() {
    return this._justified;
  };

  @Input() private get type() {
    return this._type;
  };

  private set vertical(value) {
    this._vertical = value;
    this.setClassMap();
  }

  private set justified(value) {
    this._justified = value;
    this.setClassMap();
  }

  private set type(value) {
    this._type = value;
    this.setClassMap();
  }

  private setClassMap() {
    this.classMap = {
      'nav-stacked': this.vertical,
      'nav-justified': this.justified,
      ['nav-' + (this.type || 'tabs')]: true
    };
  }

  public tabs:Array<Tab> = [];

  private _vertical:boolean;
  private _justified:boolean;
  private _type:string;
  private classMap:any = {};

  constructor() {
  }

  ngOnInit() {
    this.type = this.type !== 'undefined' ? this.type : 'tabs';
  }

  public addTab(tab:Tab) {
    this.tabs.push(tab);
    tab.active = this.tabs.length === 1 && tab.active !== false;
  }

  public removeTab(tab:Tab) {
    let index = this.tabs.indexOf(tab);
    if (index === -1) {
      return;
    }
    // Select a new tab if the tab to be removed is selected and not destroyed
    if (tab.active && this.tabs.length > 1) {
      // If this is the last tab, select the previous tab. else, the next tab.
      let newActiveIndex = index === this.tabs.length - 1 ? index - 1 : index + 1;
      this.tabs[newActiveIndex].active = true;
    }

    this.tabs.slice(index, 1);
  }
}

// TODO: templateUrl?
@Directive({selector: 'tab, [tab]'})
export class Tab implements OnInit, OnDestroy, DoCheck {
  @Input() public heading:string;
  @Input() public disabled:boolean;

  /** tab active state toogle */
  @HostBinding('class.active')
  @Input() public get active() {
    return this._active;
  }

  @Output() public select:EventEmitter<Tab> = new EventEmitter();
  @Output() public deselect:EventEmitter<Tab> = new EventEmitter();


  public set active(active) {
    if (this.disabled && active || !active) {
      if (!active) {
        this._active = active;
      }

      this.deselect.emit(this);
      return;
    }

    this._active = active;
    this.select.emit(this);
    this.tabset.tabs.forEach((tab:Tab) => {
      if (tab !== this) {
        tab.active = false;
      }
    });
  }

  @HostBinding('class.tab-pane') private addClass = true;

  private _active:boolean;
  public headingRef:TemplateRef;

  constructor(public tabset:Tabset) {
    this.tabset.addTab(this);
  }

  ngDoCheck():boolean {
    return true;
  }

  ngOnInit() {
  }

  ngOnDestroy() {
    this.tabset.removeTab(this);
  }
}

@Directive({selector: '[tab-heading]'})
export class TabHeading {
  constructor(public templateRef:TemplateRef, tab:Tab) {
    tab.headingRef = templateRef;
  }
}

export const TAB_DIRECTIVES:Array<any> = [Tab, TabHeading, Tabset];
/**
 * @deprecated - use TAB_DIRECTIVES instead
 * @type {Tab|TabHeading|Tabset[]}
 */
export const tabs:Array<any> = [Tab, TabHeading, Tabset];
